File: scripts/zigux/check-phase3-export-uapi-layout-route.py
#!/usr/bin/env python3
"""Validate the focused Phase 3 export/UAPI layout build routes."""

from __future__ import annotations

import argparse
import tempfile
from pathlib import Path


TESTS_BUILD_PATH = Path("zigux/tests/build.zig")
LAYOUT_BUILD_PATH = Path("zigux/tests/phase3_export_uapi_layout_build.zig")

SHARED_FUNCTION = "fn addPhase3ExportUapiLayout("
NEXT_FUNCTION = "\nfn addPhase3LowLevelWrappers("
HEADER_MODULE = 'const header_family_binding = b.createModule(.{'
HEADER_IMPORT = 'root_module.addImport("header_family_binding", header_family_binding);'
EXPORT_IMPORT = 'root_module.addImport("export_shim", export_shim);'
SHARED_STEP = '"phase3-export-uapi-layout"'
DEDICATED_STEP = '"phase3-export-uapi-layout-test"'


def _read(path: Path) -> str:
    return path.read_text(encoding="utf-8")


def _write(path: Path, text: str) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(text, encoding="utf-8", newline="\n")


def _extract_shared_layout_function(text: str) -> str | None:
    start = text.find(SHARED_FUNCTION)
    if start == -1:
        return None
    end = text.find(NEXT_FUNCTION, start)
    if end == -1:
        return text[start:]
    return text[start:end]


def validate_repo(repo_root: Path) -> list[str]:
    issues: list[str] = []

    tests_build = repo_root / TESTS_BUILD_PATH
    try:
        shared_text = _read(tests_build)
    except FileNotFoundError:
        return [f"missing repo file: {TESTS_BUILD_PATH.as_posix()}"]

    shared_function = _extract_shared_layout_function(shared_text)
    if shared_function is None:
        issues.append(
            f"missing {TESTS_BUILD_PATH.as_posix()} function: addPhase3ExportUapiLayout"
        )
    else:
        for marker in (HEADER_MODULE, HEADER_IMPORT, EXPORT_IMPORT, SHARED_STEP):
            if marker not in shared_function:
                issues.append(
                    "missing scoped export/UAPI layout route marker in "
                    f"{TESTS_BUILD_PATH.as_posix()}: {marker}"
                )

    dedicated_build = repo_root / LAYOUT_BUILD_PATH
    try:
        dedicated_text = _read(dedicated_build)
    except FileNotFoundError:
        issues.append(f"missing repo file: {LAYOUT_BUILD_PATH.as_posix()}")
    else:
        for marker in (HEADER_MODULE, HEADER_IMPORT, EXPORT_IMPORT, DEDICATED_STEP):
            if marker not in dedicated_text:
                issues.append(
                    f"missing dedicated export/UAPI layout marker in {LAYOUT_BUILD_PATH.as_posix()}: {marker}"
                )

    return issues


def run_self_test() -> int:
    with tempfile.TemporaryDirectory(prefix="zigux_phase3_export_uapi_layout_route_") as temp_dir:
        root = Path(temp_dir)
        shared_text = f"""
const std = @import("std");

fn addPhase3ExportUapiLayout() void {{
    {HEADER_MODULE}
    }});
    {HEADER_IMPORT}
    {EXPORT_IMPORT}
    const step = {SHARED_STEP};
}}

fn addPhase3LowLevelWrappers() void {{}}
"""
        dedicated_text = f"""
const std = @import("std");

pub fn build() void {{
    {HEADER_MODULE}
    }});
    {HEADER_IMPORT}
    {EXPORT_IMPORT}
    const step = {DEDICATED_STEP};
}}
"""

        _write(root / TESTS_BUILD_PATH, shared_text)
        _write(root / LAYOUT_BUILD_PATH, dedicated_text)
        issues = validate_repo(root)
        if issues:
            print("PHASE3_EXPORT_UAPI_LAYOUT_ROUTE_SELF_TEST=fail")
            print("\n".join(issues))
            return 1

        _write(root / TESTS_BUILD_PATH, shared_text.replace(HEADER_IMPORT, "", 1))
        issues = validate_repo(root)
        expected = (
            "missing scoped export/UAPI layout route marker in "
            f"{TESTS_BUILD_PATH.as_posix()}: {HEADER_IMPORT}"
        )
        if expected not in issues:
            print("PHASE3_EXPORT_UAPI_LAYOUT_ROUTE_SELF_TEST=fail")
            print("expected missing shared scoped header-family import to fail validation")
            return 1

        _write(root / TESTS_BUILD_PATH, shared_text.replace(NEXT_FUNCTION, "", 1))
        _write(root / LAYOUT_BUILD_PATH, dedicated_text.replace(HEADER_IMPORT, "", 1))
        issues = validate_repo(root)
        expected = (
            f"missing dedicated export/UAPI layout marker in {LAYOUT_BUILD_PATH.as_posix()}: "
            f"{HEADER_IMPORT}"
        )
        if expected not in issues:
            print("PHASE3_EXPORT_UAPI_LAYOUT_ROUTE_SELF_TEST=fail")
            print("expected missing dedicated header-family import to fail validation")
            return 1

    print("PHASE3_EXPORT_UAPI_LAYOUT_ROUTE_SELF_TEST=pass")
    print("PHASE3_EXPORT_UAPI_LAYOUT_ROUTE_SELF_TEST_CASES=3")
    return 0


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Validate the focused Phase 3 export/UAPI layout build routes."
    )
    parser.add_argument(
        "--repo-root",
        type=Path,
        default=Path("."),
        help="repository root that contains zigux/tests/",
    )
    parser.add_argument("--self-test", action="store_true")
    args = parser.parse_args()

    if args.self_test:
        return run_self_test()

    issues = validate_repo(args.repo_root)
    if issues:
        print("PHASE3_EXPORT_UAPI_LAYOUT_ROUTE=fail")
        for issue in issues:
            print(issue)
        return 1

    print(f"validated {args.repo_root / TESTS_BUILD_PATH}")
    print(f"validated {args.repo_root / LAYOUT_BUILD_PATH}")
    print("PHASE3_EXPORT_UAPI_LAYOUT_ROUTE=pass")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/zigux/check_phase3_export_uapi_layout_route.zig
const std = @import("std");
const Io = std.Io;
const guard = @import("zigux_guard.zig");

pub const live_pass_marker = "PHASE3_EXPORT_UAPI_LAYOUT_ROUTE=pass";
pub const self_test_pass_marker = "PHASE3_EXPORT_UAPI_LAYOUT_ROUTE_SELF_TEST=pass";

const SHARED_FUNCTION = [_][]const u8{
    "fn addPhase3ExportUapiLayout(",
};

const NEXT_FUNCTION = [_][]const u8{
    "\nfn addPhase3LowLevelWrappers(",
};

const HEADER_MODULE = [_][]const u8{
    "const header_family_binding = b.createModule(.{",
};

const HEADER_IMPORT = [_][]const u8{
    "root_module.addImport(\"header_family_binding\", header_family_binding);",
};

const EXPORT_IMPORT = [_][]const u8{
    "root_module.addImport(\"export_shim\", export_shim);",
};

const SHARED_STEP = [_][]const u8{
    "\"phase3-export-uapi-layout\"",
};

const DEDICATED_STEP = [_][]const u8{
    "\"phase3-export-uapi-layout-test\"",
};

fn checkRepo(io: Io, allocator: std.mem.Allocator, root: []const u8) !void {
    const text_shared_function_path = try guard.joinPath(allocator, root, "zigux/tests/build.zig");
    defer allocator.free(text_shared_function_path);
    const text_shared_function = try guard.readUtf8File(io, allocator, text_shared_function_path);
    defer allocator.free(text_shared_function);
    for (SHARED_FUNCTION) |marker| try guard.requireMarker(text_shared_function, marker);
    const text_next_function_path = try guard.joinPath(allocator, root, "zigux/tests/build.zig");
    defer allocator.free(text_next_function_path);
    const text_next_function = try guard.readUtf8File(io, allocator, text_next_function_path);
    defer allocator.free(text_next_function);
    for (NEXT_FUNCTION) |marker| try guard.requireMarker(text_next_function, marker);
    const text_header_module_path = try guard.joinPath(allocator, root, "zigux/tests/build.zig");
    defer allocator.free(text_header_module_path);
    const text_header_module = try guard.readUtf8File(io, allocator, text_header_module_path);
    defer allocator.free(text_header_module);
    for (HEADER_MODULE) |marker| try guard.requireMarker(text_header_module, marker);
    const text_header_import_path = try guard.joinPath(allocator, root, "zigux/tests/build.zig");
    defer allocator.free(text_header_import_path);
    const text_header_import = try guard.readUtf8File(io, allocator, text_header_import_path);
    defer allocator.free(text_header_import);
    for (HEADER_IMPORT) |marker| try guard.requireMarker(text_header_import, marker);
    const text_export_import_path = try guard.joinPath(allocator, root, "zigux/tests/build.zig");
    defer allocator.free(text_export_import_path);
    const text_export_import = try guard.readUtf8File(io, allocator, text_export_import_path);
    defer allocator.free(text_export_import);
    for (EXPORT_IMPORT) |marker| try guard.requireMarker(text_export_import, marker);
    const text_shared_step_path = try guard.joinPath(allocator, root, "zigux/tests/build.zig");
    defer allocator.free(text_shared_step_path);
    const text_shared_step = try guard.readUtf8File(io, allocator, text_shared_step_path);
    defer allocator.free(text_shared_step);
    for (SHARED_STEP) |marker| try guard.requireMarker(text_shared_step, marker);
    const text_dedicated_step_path = try guard.joinPath(allocator, root, "zigux/tests/build.zig");
    defer allocator.free(text_dedicated_step_path);
    const text_dedicated_step = try guard.readUtf8File(io, allocator, text_dedicated_step_path);
    defer allocator.free(text_dedicated_step);
    for (DEDICATED_STEP) |marker| try guard.requireMarker(text_dedicated_step, marker);
}

fn runSelfTest(io: Io, allocator: std.mem.Allocator) !u8 {
    try checkRepo(io, allocator, try guard.defaultRepoRoot(allocator));
    try guard.printLine(io, "{s}", .{self_test_pass_marker});
    return 0;
}

pub fn main(init: std.process.Init) !void {
    const allocator = init.gpa;
    const io = init.io;
    const args = try init.minimal.args.toSlice(allocator);

    var self_test = false;
    var explicit_root: ?[]const u8 = null;
    var index: usize = 1;
    while (index < args.len) : (index += 1) {
        const arg = args[index];
        if (std.mem.eql(u8, arg, "--self-test")) {
            self_test = true;
            continue;
        }
        if (std.mem.eql(u8, arg, "--root")) {
            if (index + 1 >= args.len) std.process.exit(2);
            index += 1;
            explicit_root = args[index];
            continue;
        }
    }

    const root = explicit_root orelse try guard.repoRootFromScript(allocator);
    defer if (explicit_root == null) allocator.free(root);

    if (self_test) {
        std.process.exit(try runSelfTest(io, allocator));
    }

    checkRepo(io, allocator, root) catch {
        std.process.exit(1);
    };
    try guard.printLine(io, "{s}", .{live_pass_marker});
}
